src/app-layer-detect-proto.c

```diff
@@ -1612,6 +1612,24 @@ void AppLayerProtoDetectRegisterProtocol(AppProto alproto, char *alproto_name)
     SCReturn;
 }
 
+void AppLayerProtoDetectReset(Flow *f)
+{
+    FlowUnsetChangeProtoFlag(f);
+    FLOW_RESET_PM_DONE(f, STREAM_TOSERVER);
+    FLOW_RESET_PM_DONE(f, STREAM_TOCLIENT);
+    FLOW_RESET_PP_DONE(f, STREAM_TOSERVER);
+    FLOW_RESET_PP_DONE(f, STREAM_TOCLIENT);
+    f->probing_parser_toserver_alproto_masks = 0;
+    f->probing_parser_toclient_alproto_masks = 0;
+    f->alproto    = ALPROTO_UNKNOWN;
+    f->alproto_ts = ALPROTO_UNKNOWN;
+    f->alproto_tc = ALPROTO_UNKNOWN;
+
+    AppLayerParserStateCleanup(f->proto, f->alproto, f->alstate, f->alparser);
```

Comment:
The reason for the memory leak is that f->alproto is set to ALPROTO_UNKNOWN before AppLayerParserStateCleanup is called.